### PR TITLE
[Feature] [Connector-V2] Add connector amazonsqs

### PR DESCRIPTION
**Pull request by HaoXuAI**
## Purpose of this pull request

Add new connector amazon sqs (https://aws.amazon.com/sqs/)

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by liugddx** [comment]
Please add e2e test.

**Comment by HaoXuAI** [comment]
> Please add e2e test.

I'm working on it. Unfortunately, SQS doesn't have a local service to test. My team use https://localstack.cloud/ to mock the service. Is it ok with you to leverage it in ST too? @liugddx

**Comment by liugddx** [comment]
> > Please add e2e test.
> 
> I'm working on it. Unfortunately, SQS doesn't have a local service to test. My team use https://localstack.cloud/ to mock the service. Is it ok with you to leverage it in ST too? @liugddx

You can write e2e and Add @disabled annotation.

File: .gitignore [hunk 1]
```diff
@@ -48,4 +48,7 @@ test.conf
 spark-warehouse
 *.flattened-pom.xml
 
-seatunnel-examples
\ No newline at end of file
+seatunnel-examples
+
+# vscode
+.vscode
```
Comment (zhilinli123) [inline .gitignore:54]: Why add it?

Comment (zhilinli123) [inline .gitignore:54]: Why add it?

Comment (HaoXuAI) [reply .gitignore:54]: to avoid uploading my own vscode settings.

**Comment by zhilinli123** [comment]
HI @HaoXuAI At present, we use the V2.0 template design of Connector Doc for your reference,You may need to make some improvements to the documentation. Thank you
https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/source/Mysql.md
issues:https://github.com/apache/seatunnel/issues/4572 CC @liugddx

**Comment by HaoXuAI** [comment]
> HI @HaoXuAI At present, we use the V2.0 template design of Connector Doc for your reference,You may need to make some improvements to the documentation. Thank you https://github.com/apache/seatunnel/blob/dev/docs/en/connector-v2/source/Mysql.md issues:#4572 CC @liugddx

thanks. I have updated the docs.

**Comment by HaoXuAI** [comment]
> > > Please add e2e test.
> > 
> > 
> > I'm working on it. Unfortunately, SQS doesn't have a local service to test. My team use https://localstack.cloud/ to mock the service. Is it ok with you to leverage it in ST too? @liugddx
> 
> You can write e2e and Add @disabled annotation.

I have added the e2e test and it passed.

**Review by FlechazoW — APPROVED** [review]
Nice job. LGTM.

**Review by liugddx — DISMISSED** [review]
LGTM

File: docs/en/connector-v2/source/AmazonSqs.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+transform {
+    # If you would like to get more information about how to configure seatunnel and see full list of transform plugins,
+    # please go to https://seatunnel.apache.org/docs/transform-v2/sql
+}
+
+sink {
+    Console {}
+}
```
Comment (zhilinli123) [inline docs/en/connector-v2/source/AmazonSqs.md:73]: I think this demo should be written to amazonsqs as an example

File: docs/en/connector-v2/source/AmazonSqs.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+| field_delimiter         | String | No       | ,       | Customize the field delimiter for data format.                                                                                                                                                                                                                                                                                                                                                                                      |
+| common-options          |        | No       | -       | Source plugin common parameters, please refer to [Source Common Options](common-options.md) for details                                                                                                                                                                                                                                                                                                                             |
+
+## Task Example
+
+```bash
+source {
```
Comment (zhilinli123) [inline docs/en/connector-v2/source/AmazonSqs.md:39]: Add a format source demo?

File: docs/en/connector-v2/sink/AmazonSqs.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+```bash
+amazonsqs {
+    url = "http://127.0.0.1:8000"
+    region = "us-east-1"
+    queue = "queueName"
+    format = json
+  }
+```
+
```
Comment (zhilinli123) [inline docs/en/connector-v2/sink/AmazonSqs.md]: Full case conf

File: docs/en/connector-v2/sink/AmazonSqs.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+| field_delimiter         | String | No       | ,       | Customize the field delimiter for data format.                                                                                                                                                                                                                                                                                                                                                                                      |
+| common-options          |        | No       | -       | Source plugin common parameters, please refer to [Source Common Options](common-options.md) for details                                                                                                                                                                                                                                                                                                                             |
+
+## Example
```
Comment (zhilinli123) [inline docs/en/connector-v2/sink/AmazonSqs.md]: Example-> Task Example

File: docs/en/connector-v2/sink/AmazonSqs.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
+# AmazonSqs
+
+> Amazon SQS sink connector
+
+## Support Those Engines
```
Comment (zhilinli123) [inline docs/en/connector-v2/sink/AmazonSqs.md:5]: Add titles for supported versions : `localstack/localstack:0.11.2`  For user reference source&sink doc

File: docs/en/connector-v2/sink/AmazonSqs.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+## Key features
+
+- [x] [batch](../../concept/connector-v2-features.md)
+- [x] [stream](../../concept/connector-v2-features.md)
+- [x] [exactly-once](../../concept/connector-v2-features.md)
+- [ ] [column projection](../../concept/connector-v2-features.md)
+- [ ] [parallelism](../../concept/connector-v2-features.md)
+- [ ] [support user-defined split](../../concept/connector-v2-features.md)
+
+## Options
```
Comment (zhilinli123) [inline docs/en/connector-v2/sink/AmazonSqs.md]: amazonsqs and SeaTunnel type mapping We also want to tell users that it is best to reflect this in the documentation source&sink doc

File: docs/en/connector-v2/source/AmazonSqs.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
+# AmazonSqs
+
+> AmazonSqs source connector
+
```
Comment (zhilinli123) [inline docs/en/connector-v2/source/AmazonSqs.md:4]: Add a supported version as above

File: docs/en/connector-v2/source/AmazonSqs.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+- [x] [stream](../../concept/connector-v2-features.md)
+- [ ] [exactly-once](../../concept/connector-v2-features.md)
+- [ ] [column projection](../../concept/connector-v2-features.md)
+- [ ] [parallelism](../../concept/connector-v2-features.md)
+- [ ] [support user-defined split](../../concept/connector-v2-features.md)
+
+## Description
+
+Read data from Amazon SQS.
+
```
Comment (zhilinli123) [inline docs/en/connector-v2/source/AmazonSqs.md:23]: Add type mapping as above

File: docs/en/connector-v2/sink/AmazonSqs.md (hunk 7 above)
Comment (HaoXuAI) [reply docs/en/connector-v2/sink/AmazonSqs.md]: the sqs is using message ( json/text) doesn't have a type, so we don't need type mapping i believe

File: docs/en/connector-v2/sink/AmazonSqs.md (hunk 6 above)
Comment (HaoXuAI) [reply docs/en/connector-v2/sink/AmazonSqs.md:5]: localstack is not a engine but only used for testing purpose.

**Comment by HaoXuAI** [comment]
@zhilinli123 updated the doc, let me know if you have other comments.

File: docs/en/connector-v2/sink/AmazonSqs.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
```
Comment (zhilinli123) [inline docs/en/connector-v2/sink/AmazonSqs.md]: Sink Options

File: docs/en/connector-v2/sink/AmazonSqs.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+> Spark<br/>
+> Flink<br/>
+> SeaTunnel Zeta<br/>
+
+## Description
+
+Write data to Amazon SQS
+
+## Key features
```
Comment (zhilinli123) [inline docs/en/connector-v2/sink/AmazonSqs.md]: ## Key Features

File: docs/en/connector-v2/sink/AmazonSqs.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+## Task Example
+
+```bash
+amazonsqs {
+    url = "http://127.0.0.1:8000"
+    region = "us-east-1"
+    queue = "queueName"
+    format = text
+    field_delimiter = "|"
+
```
Comment (zhilinli123) [inline docs/en/connector-v2/sink/AmazonSqs.md]: 1.Delete line feed
2.add Source case
```
source {
  FakeSource {
    schema = {
      fields {
        c_map = "map<string, string>"
        c_array = "array<int>"
        c_string = string
        c_boolean = boolean
        c_tinyint = tinyint
        c_smallint = smallint
        c_int = int
        c_bigint = bigint
        c_float = float
        c_double = double
        c_bytes = bytes
        c_date = date
        c_decimal = "decimal(38, 18)"
        c_timestamp = timestamp
        c_row = {
          c_map = "map<string, string>"
          c_array = "array<int>"
          c_string = string
          c_boolean = boolean
          c_tinyint = tinyint
          c_smallint = smallint
          c_int = int
          c_bigint = bigint
          c_float = float
          c_double = double
          c_bytes = bytes
          c_date = date
          c_decimal = "decimal(38, 18)"
          c_timestamp = timestamp
        }
      }
    }
    result_table_name = "fake"
  }
}
```

File: docs/en/connector-v2/source/AmazonSqs.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+> AmazonSqs source connector
+
+## Support Those Engines
+
+> Spark<br/>
+> Flink<br/>
+> SeaTunnel Zeta<br/>
+
+## Key features
```
Comment (zhilinli123) [inline docs/en/connector-v2/source/AmazonSqs.md]: Key Features

**Comment by zhilinli123** [comment]
I add a few small suggestions, others look good
@HaoXuAI

**Comment by HaoXuAI** [comment]
> I add a few small suggestions, others look good @HaoXuAI

Thanks! Updated.

**Comment by zhilinli123** [comment]
wait CI good 

> > I add a few small suggestions, others look good @HaoXuAI
> 
> Thanks! Updated.

wait ci looks good

**Review by liugddx — DISMISSED** [review]

**Review by zhilinli123 — APPROVED** [review]

File: seatunnel-connectors-v2/connector-amazonsqs/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazonsqs/source/AmazonSqsSourceFactory.java [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+@AutoService(Factory.class)
+public class AmazonSqsSourceFactory implements TableSourceFactory {
+    @Override
+    public String factoryIdentifier() {
+        return "AmazonSqs";
+    }
+
+    @Override
+    public OptionRule optionRule() {
+        return OptionRule.builder().required(URL, REGION, CatalogTableUtil.SCHEMA).build();
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-amazonsqs/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazonsqs/source/AmazonSqsSourceFactory.java]: Please rebase from dev, I believe this field already move to `TableSchemaOptions`

File: seatunnel-connectors-v2/connector-amazonsqs/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazonsqs/config/AmazonSqsConfig.java [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription("The access secret key of Amazon SQS Service");
+
+    public static final Option<String> MESSAGE_GROUP_ID =
+            Options.key("message_group_id")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription("The message group id of Amazon SQS Service");
+    public static final Option<String> SCHEMA =
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-amazonsqs/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazonsqs/config/AmazonSqsConfig.java]: Please reuse `TableSchemaOptions.SCHEMA`, do not create new `schema` option.

Comment (HaoXuAI) [reply seatunnel-connectors-v2/connector-amazonsqs/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazonsqs/config/AmazonSqsConfig.java]: updated. thanks!

File: seatunnel-connectors-v2/connector-amazonsqs/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazonsqs/sink/AmazonSqsSinkFactory.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+import org.apache.seatunnel.api.table.factory.Factory;
+import org.apache.seatunnel.api.table.factory.TableSinkFactory;
+
+import com.google.auto.service.AutoService;
+
+import static org.apache.seatunnel.connectors.seatunnel.amazonsqs.config.AmazonSqsConfig.REGION;
+import static org.apache.seatunnel.connectors.seatunnel.amazonsqs.config.AmazonSqsConfig.URL;
+
+@AutoService(Factory.class)
+public class AmazonSqsSinkFactory implements TableSinkFactory {
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-amazonsqs/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazonsqs/sink/AmazonSqsSinkFactory.java:34]: Could you implement `TabkeSinkFactory::createSink` to create sink instance? We will remove `SeaTunnelSink::prepare` method in the future. You can refer `JdbcSinkFactory`. Please.

Comment (HaoXuAI) [reply seatunnel-connectors-v2/connector-amazonsqs/src/main/java/org/apache/seatunnel/connectors/seatunnel/amazonsqs/sink/AmazonSqsSinkFactory.java:34]: @Hisoka-X sounds good and updated. thanks

**Comment by Hisoka-X** [comment]
The test case seem like not passed. Could you fix it? Thanks.

**Comment by HaoXuAI** [comment]
> The test case seem like not passed. Could you fix it? Thanks.

@Hisoka-X the error is not clear to me: https://github.com/HaoXuAI/incubator-seatunnel/actions/runs/6570589755/job/17848343358, can you help to take a look of it. thanks!

**Comment by Hisoka-X** [comment]
I believe the error came from constructor
```
Caused by: org.apache.seatunnel.common.utils.SeaTunnelException: java.lang.NullPointerException
	at org.apache.seatunnel.connectors.seatunnel.amazonsqs.sink.AmazonSqsSinkWriter.<init>(AmazonSqsSinkWriter.java:61)
	at org.apache.seatunnel.connectors.seatunnel.amazonsqs.sink.AmazonSqsSink.createWriter(AmazonSqsSink.java:90)
	at org.apache.seatunnel.connectors.seatunnel.amazonsqs.sink.AmazonSqsSink.createWriter(AmazonSqsSink.java:46)
	at org.apache.seatunnel.engine.server.task.flow.SinkFlowLifeCycle.restoreState(SinkFlowLifeCycle.java:286)
	at org.apache.seatunnel.engine.server.task.SeaTunnelTask.lambda$restoreState$15(SeaTunnelTask.java:419)
```

**Comment by HaoXuAI** [comment]
> I believe the error came from constructor
> 
> ```
> Caused by: org.apache.seatunnel.common.utils.SeaTunnelException: java.lang.NullPointerException
> 	at org.apache.seatunnel.connectors.seatunnel.amazonsqs.sink.AmazonSqsSinkWriter.<init>(AmazonSqsSinkWriter.java:61)
> 	at org.apache.seatunnel.connectors.seatunnel.amazonsqs.sink.AmazonSqsSink.createWriter(AmazonSqsSink.java:90)
> 	at org.apache.seatunnel.connectors.seatunnel.amazonsqs.sink.AmazonSqsSink.createWriter(AmazonSqsSink.java:46)
> 	at org.apache.seatunnel.engine.server.task.flow.SinkFlowLifeCycle.restoreState(SinkFlowLifeCycle.java:286)
> 	at org.apache.seatunnel.engine.server.task.SeaTunnelTask.lambda$restoreState$15(SeaTunnelTask.java:419)
> ```

@Hisoka-X thanks, I've fixed the issue.
btw looks like the `seatunnel-example` won't use the `sinkfactory` to create sink, so that's not able to test via just the java example code.

**Review by hailin0 — APPROVED** [review]
LGTM

**Review by Hisoka-X — APPROVED** [review]
Thanks @HaoXuAI !

**Pull request merged**